src/box/lua/config/applier/box_cfg.lua

```diff
@@ -283,11 +283,26 @@ local function apply(config)
     -- mix instances with data from different replicasets into one
     -- replicaset.
     local names = configdata:names()
-    box_cfg.instance_name = names.instance_name
-    box_cfg.replicaset_name = names.replicaset_name
     box_cfg.instance_uuid = names.instance_uuid
     box_cfg.replicaset_uuid = names.replicaset_uuid
 
+    -- Names are applied only if they're already in snap file.
+    -- Otherwise, master must apply names after box.cfg.
+    if not configdata._missing_names[names.instance_name] then
+        box_cfg.instance_name = names.instance_name
+    end
+    if not configdata._missing_names[names.replicaset_name] then
+        box_cfg.replicaset_name = names.replicaset_name
+    end
+
+    if is_startup then
+        for name, uuid in pairs(configdata._missing_names) do
+            local warning = string.format('box_cfg.apply: name %s for '..
+                                          '%s is missing', name, uuid)
+            config:_alert(name, {type = 'warn', message = warning})
+        end
+    end
```

Comment:
AFAIU, if we react on new instances in the configuration only on startup, we don't show an alert if an instance that is added into the config has no instance name in the database.

Reply:
From now on we recreate `missing_names` every time, `configdata` is created. So, we should make these alerts every time, box.cfg is called.